# Homework 5
## EEEL 4220 Fall 2022

<font color='blue'>**Final answers in blue**</font>

### Problem 1 KKT conditoins (30)
Consider the following optimization problem
$$
\begin{gather}
\min_{x,y} x^2 + 2x + 2y^2 + 5y + z\\
\text{s.t.}\\
2x + z \geq 0 \\
x + y + z = 5
\end{gather}
$$

1. Write out the Lagrangian function (5)
2. List the KKT condition (5)
3. Solve the optimization problem using the KKT condition (20)

**Note:** You can use CVXPY to check your result but do not write code as answer to this queston.

Clarification on Ed: need to minimize with respect to x,y,z not just x and y

<font color='blue'>**Start of Problem 1 Part 1**</font>

Solution using Lagrangian multipliers from Lecture 4 Slide 89:
$$\ell(x_1,...,x_n,\lambda_1,...,\lambda_m,\mu_1,...,\mu_p) = f(x_1,...,x_n) + \Sigma_{i=1}^m \lambda_i*\omega(x_1,..,x_n) + \Sigma_{j=1}^p \mu_j*g_j(x_1,..,x_n)$$

For this case:

$$\ell(x,y,z,\lambda_1,...,\lambda_m,\mu_1,...,\mu_p) = f(x,y,z) + \Sigma_{i=1}^m \lambda_i*\omega(x,y,z) + \Sigma_{j=1}^p \mu_j*g_j(x,y,z)$$

Function to be minimized:

$$ f(x,y,z) = x^2 + 2x + 2y^2 + 5y + z $$

Equality constraint $\omega$: 

$$\omega = 5 - x - y - z $$

Also given: 

$$ 2x + z \geq 0 $$

Rearranging for inequality constraint $g$:
$$  -2x - z \leq 0 $$

Just have the 1 equality constraint $\omega$, thus only 1 $\lambda$ or $m$=1:

$$\ell(x,y,z,\lambda,\mu_1,...,\mu_p) = f(x,y,z) + \lambda*\omega(x,y,z) + \Sigma_{j=1}^p \mu_j*g_j(x,y,z)$$

Just have the 1 inequality constraint $g$, thus only 1 $\mu$ or $p$=1:

$$\ell(x,y,z,\lambda,\mu) = f(x,y,z) + \lambda*\omega(x,y,z) + \mu*g(x,y,z)$$

<font color='blue'>Thus Lagrangian is:

<font color='blue'>$$\ell(x,y,z,\lambda,\mu) = x^2 + 2x + 2y^2 + 5y + z + \lambda*[5 - x - y - z] + \mu*[-2x - z]$$

<font color='blue'>**Answer to Problem 1 Part 1**</font>

<font color='blue'>$$\ell(x,y,z,\lambda,\mu) = x^2 + 2x + 2y^2 + 5y + z + \lambda*[5 - x - y - z] + \mu*[-2x - z]$$

<font color='blue'>**Start of Problem 1 Part 2**</font>

from slide 88 of Lecture 4:

$$\ell(x,\lambda,\mu) = f(x) + \Sigma_{i=1}^m \lambda_i*\omega_i(x) + \Sigma_{j=1}^p \mu_j*g_j(x)$$

Karush Kuhn Tucker (KKT) condition (also listed on same slide):

$$\frac{\partial\ell(x,\lambda,\mu)}{\partial x_i} \equiv \frac{\partial f(x)}{\partial x_i} +  \Sigma_{i=1}^m \lambda_i *\frac{\partial \omega_i(x)}{\partial x_i} + \Sigma_{j=1}^p \mu_j* \frac{\partial g_j(x)}{\partial x_i} = 0$$


$$\frac{\partial\ell(x,\lambda,\mu)}{\partial \lambda_k} \equiv \omega_k(x) = 0$$

$$ g_j(x) \leq 0$$

With complementary slackness conditions:

$$ \mu_j g_j(x) = 0 $$
$$ \mu_j \geq 0 $$

Thus KKT condition is as follows:

$$\frac{\partial\ell(x,\lambda,\mu)}{\partial x_i} \equiv \frac{\partial [x^2 + 2x + 2y^2 + 5y + z]}{\partial x_i} +  \lambda *\frac{\partial [5 - x - y - z]}{\partial x_i} + \mu* \frac{\partial [ -2x - z ]}{\partial x_i} = 0$$

<font color='blue'>$$\frac{\partial\ell(x,\lambda,\mu)}{\partial \lambda} \equiv 5 - x - y - z = 0$$

<font color='blue'>$$ g_j(x)\equiv -2x - z \leq 0$$

With respect to x:

$$\frac{\partial\ell(x,\lambda,\mu)}{\partial x} \equiv \frac{\partial [x^2 + 2x + 2y^2 + 5y + z]}{\partial x} +  \lambda *\frac{\partial [5 - x - y - z]}{\partial x} + \mu* \frac{\partial [ -2x - z ]}{\partial x} = 0$$

$$\frac{\partial\ell(x,\lambda,\mu)}{\partial x} \equiv [2x + 2] +  \lambda *[-1] + \mu*[ -2] = 0$$

<font color='blue'>$$\frac{\partial\ell(x,\lambda,\mu)}{\partial x} \equiv 2x + 2 - \lambda - 2 \mu = 0$$

With respect to y:

$$\frac{\partial\ell(x,\lambda,\mu)}{\partial y} \equiv \frac{\partial [x^2 + 2x + 2y^2 + 5y + z]}{\partial y} +  \lambda *\frac{\partial [5 - x - y - z]}{\partial y} + \mu* \frac{\partial [ -2x - z ]}{\partial y} = 0$$

$$\frac{\partial\ell(x,\lambda,\mu)}{\partial y} \equiv [2*2y + 5] +  \lambda *[-1] + \mu* [0] = 0$$

<font color='blue'>$$\frac{\partial\ell(x,\lambda,\mu)}{\partial y} \equiv 4y + 5 - \lambda = 0$$

With respect to z:

$$\frac{\partial\ell(x,\lambda,\mu)}{\partial z} \equiv \frac{\partial [x^2 + 2x + 2y^2 + 5y + z]}{\partial z} +  \lambda *\frac{\partial [5 - x - y - z]}{\partial z} + \mu* \frac{\partial [ -2x - z ]}{\partial z} = 0$$

$$\frac{\partial\ell(x,\lambda,\mu)}{\partial z} \equiv [1] +  \lambda * [-1]  + \mu* [-1] = 0$$

<font color='blue'>$$\frac{\partial\ell(x,\lambda,\mu)}{\partial z} \equiv 1 -  \lambda  - \mu = 0$$

Plugging into complementary slackness conditions:

$$ \mu g(x) = \mu[-2x - z] = 0 $$

<font color='blue'>$$ \mu g(x) = -2x\mu - z\mu = 0 $$

<font color='blue'>$$ \mu \geq 0 $$

<font color='blue'>**Answer to Problem 1 Part 2**</font>

<font color='blue'>**In summary, KKT condition is as follows:**

<font color='blue'>$$\frac{\partial\ell(x,\lambda,\mu)}{\partial x} \equiv 2x + 2 - \lambda - 2 \mu = 0$$

<font color='blue'>$$\frac{\partial\ell(x,\lambda,\mu)}{\partial y} \equiv 4y + 5 - \lambda = 0$$

<font color='blue'>$$\frac{\partial\ell(x,\lambda,\mu)}{\partial z} \equiv 1 -  \lambda  - \mu = 0$$

<font color='blue'>$$\frac{\partial\ell(x,\lambda,\mu)}{\partial \lambda} \equiv 5 - x - y - z = 0$$

<font color='blue'>$$ g(x)\equiv -2x - z \leq 0$$

<font color='blue'>**With complementary slackness conditions:**

<font color='blue'>$$ \mu g(x) = -2x\mu - z\mu = 0 $$

<font color='blue'>$$ \mu \geq 0 $$

<font color='blue'>**Start of Problem 1 Part 3**</font>

First assume inequality constraint not binding (slide 94):
$$ \mu = 0 $$

Plugging into eqn wrt x:

$$\frac{\partial\ell(x,\lambda,\mu)}{\partial x} \equiv 2x + 2 - \lambda - 2 \mu = 0$$

$$\frac{\partial\ell(x,\lambda,\mu)}{\partial x} \equiv 2x + 2 - \lambda - 2 [0] = 0$$

$$\frac{\partial\ell(x,\lambda,\mu)}{\partial x} \equiv 2x + 2 - \lambda = 0$$

Plugging into eqn wrt y:
$$\frac{\partial\ell(x,\lambda,\mu)}{\partial y} \equiv 4y + 5 - \lambda = 0$$

Plugging into eqn wrt z:
$$\frac{\partial\ell(x,\lambda,\mu)}{\partial z} \equiv 1 -  \lambda  - \mu = 0$$

$$\frac{\partial\ell(x,\lambda,\mu)}{\partial z} \equiv 1 -  \lambda  - [0] = 0$$

$$\frac{\partial\ell(x,\lambda,\mu)}{\partial z} \equiv 1 -  \lambda = 0$$

Solving eqn wrt z:

$$\frac{\partial\ell(x,\lambda,\mu)}{\partial z} \equiv 1 -  \lambda = 0$$

$$ 1 -  \lambda = 0$$

$$ \lambda = 1$$

Solving eqn wrt x:

$$\frac{\partial\ell(x,\lambda,\mu)}{\partial x} \equiv 2x + 2 - \lambda = 0$$

$$ 2x + 2 - [1] = 0$$

$$ 2x + 1 = 0$$

<font color='blue'>$$ x = -0.5$$

Solving eqn wrt y:

$$\frac{\partial\ell(x,\lambda,\mu)}{\partial y} \equiv 4y + 5 - \lambda = 0$$

$$ 4y + 5 - [1] = 0$$

$$ 4y + 4 = 0$$

<font color='blue'>$$ y = -1$$

Find z:

$$x + y + z = 5$$

$$z = 5 - x - y $$

$$z = 5 - (-0.5) - (-1)$$

<font color='blue'>$$z = 6.5$$

Check feasibility:
$$  -2x - z \leq 0 ?$$

$$  -2[-0.5] - [6.5] \leq 0 ?$$

$$  1 - [6.5] \leq 0 ?$$

$$  -5.5 \leq 0 $$

Thus, solution is feasible

Find value of objective function for $x = -0.5$, $y= -1$, $z=6.5$:

$$x^2 + 2x + 2y^2 + 5y + z$$

$$(-0.5)^2 + 2(-0.5) + 2(-1)^2 + 5(-1) + (6.5)$$

$$0.25 -1 + 2 -5 + 6.5$$

$$0.25 + 1 + 1.5$$

<font color='blue'>$$ 2.75$$

<font color='blue'>**Answer to Problem 1 Part 3**</font>

<font color='blue'>$x = -0.5$, $y= -1$, $z=6.5$ for objective function value of $2.75$

In [120]:
# check with CVXPY:
import cvxpy as cp

In [121]:
# Define variables
x = cp.Variable(nonneg = False)
y = cp.Variable(nonneg = False)
z = cp.Variable(nonneg = False)

In [122]:
# Create objective function
obj = cp.Minimize(x**2 + 2*x + 2*y**2 + 5*y + z)

In [123]:
con = [];
# Define inequality constraints 
con += [-2*x - z <= 0];
# Define equality constraints 
con += [5 - x - y -z == 0]

In [124]:
# Solve the problem
prob = cp.Problem(obj, con)

In [125]:
# call solver to solve the problem
prob.solve(verbose = True) # You may specify the solver here, 
                           # otherwise ECOS is used for LPs by default

                                     CVXPY                                     
                                     v1.3.2                                    
(CVXPY) Oct 20 06:53:46 PM: Your problem has 3 variables, 2 constraints, and 0 parameters.
(CVXPY) Oct 20 06:53:46 PM: It is compliant with the following grammars: DCP, DQCP
(CVXPY) Oct 20 06:53:46 PM: (If you need to solve this problem multiple times, but with different data, consider using parameters.)
(CVXPY) Oct 20 06:53:46 PM: CVXPY will first compile your problem; then, it will invoke a numerical solver to obtain a solution.
-------------------------------------------------------------------------------
                                  Compilation                                  
-------------------------------------------------------------------------------
(CVXPY) Oct 20 06:53:46 PM: Compiling problem (target solver=OSQP).
(CVXPY) Oct 20 06:53:46 PM: Reduction chain: CvxAttr2Constr -> Qp2SymbolicQp -> QpMatrixStuffing 

2.7500000000000004

In [126]:
prob.status

'optimal'

In [127]:
#print optimal value and variables
print('\033[94mOptimal value: ', f'{prob.value:.8}' +'\033[0m')
print('\033[94mOptimal variables: ', f'x={x.value:.5}, y={y.value:.5}, z={z.value:.5}'+'\033[0m')

Optimal value:  2.75
Optimal variables:  x=-0.5, y=-1.0, z=6.5


### Problem 2 Economic dispatch using CVXPY (20)

A utility company has three generators with the following cost functions (unit in \\$/h):
$$
\begin{align}
C_1(x_1) &= 0.01 x_1^2 + 20 x_1 + 100 \\
C_2(x_2) &= 0.02 x_2^2 + 25 x_2 + 200 \\
C_3(x_3) &= 0.03 x_3^2 + 30 x_3 + 300 
\end{align}
$$

And the minimum and maximum generator capacity are listed below:
<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{border-color:black;border-style:solid;border-width:1px;font-family:Arial, sans-serif;font-size:14px;
  overflow:hidden;padding:10px 5px;word-break:normal;}
.tg th{border-color:black;border-style:solid;border-width:1px;font-family:Arial, sans-serif;font-size:14px;
  font-weight:normal;overflow:hidden;padding:10px 5px;word-break:normal;}
.tg .tg-0pky{border-color:inherit;text-align:left;vertical-align:top}
</style>
<table class="tg">
<thead>
  <tr>
    <th class="tg-0pky"> </th>
    <th class="tg-0pky">Gen Min (MW) </th>
    <th class="tg-0pky">Gem Max (MW)</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td class="tg-0pky">Generator 1</td>
    <td class="tg-0pky">100</td>
    <td class="tg-0pky">500</td>
  </tr>
  <tr>
    <td class="tg-0pky">Generator 2</td>
    <td class="tg-0pky">50</td>
    <td class="tg-0pky">300</td>
  </tr>
  <tr>
    <td class="tg-0pky">Generator 3</td>
    <td class="tg-0pky">0</td>
    <td class="tg-0pky">100</td>
  </tr>
</tbody>
</table>


We now practice to solve the optimization using CVXPY. Before we start, load the CVXPY package

In [83]:
import cvxpy as cp

First, we need to define variables:

In [84]:
# Define x1 as a variable
x1 = cp.Variable(nonneg = True)

It defines `x1` as a variable under the CVXPY class - `cp.Variable`. Since most of the time we will be dealing with nonnegative variables, it is convenient to simply enforce variables to be positive only with `(nonneg = True)`. 

1. Now define variables for generator 2 and 3 (5)

In [85]:
# Define x2 as a variable
x2 = cp.Variable(nonneg = True)
# Define x3 as a variable
x3 = cp.Variable(nonneg = True)

Next, define the objective function as a minimization problem, note that the first generator is already given.

2. Please fill in the rest two generators into the objective function (5)

In [95]:
# Fill in cost function for generator two and three
obj = cp.Minimize(
    0.01*cp.square(x1) + 20 * x1 + 100 + 
    0.02*cp.square(x2) + 25*x2 + 200 +
    0.03*cp.square(x3) + 30*x3 + 300 
                 )

Now define inequality constraints

In [108]:
con = [];
con += [x1 >= 100];
con += [x1 <= 500];

3. Define inequality constraints for generator 2 and 3 (5)

In [109]:
# Define constraints for generator 2
con += [x2 >= 50];
con += [x2 <= 300];
# Define constraints for generator 3
con += [x3 >= 0];
con += [x3 <= 100];

Finally, add the last constraint which enforces the sum of generatoin equals to the demand which is 500 MW. Note that all equality constraints must be defined using `==` instead of `=`.

4. Define the demand balance constraint (5)

In [110]:
# Define the demand balance constraint
con += [500 - x1 - x2 - x3 == 0]

So far we have defined the objective `obj` and the constraint set `con`. Now we assemble them together into a **Problem** with name `prob1`

In [111]:
# Solve the problem
prob1 = cp.Problem(obj, con)

Note we call a solver to solve `prob1` for us

In [112]:
# call solver to solve the problem
prob1.solve(verbose = True) # You may specify the solver here, 
                           # otherwise ECOS is used for LPs by default

                                     CVXPY                                     
                                     v1.3.2                                    
(CVXPY) Oct 20 06:52:59 PM: Your problem has 3 variables, 7 constraints, and 0 parameters.
(CVXPY) Oct 20 06:52:59 PM: It is compliant with the following grammars: DCP, DQCP
(CVXPY) Oct 20 06:52:59 PM: (If you need to solve this problem multiple times, but with different data, consider using parameters.)
(CVXPY) Oct 20 06:52:59 PM: CVXPY will first compile your problem; then, it will invoke a numerical solver to obtain a solution.
-------------------------------------------------------------------------------
                                  Compilation                                  
-------------------------------------------------------------------------------
(CVXPY) Oct 20 06:52:59 PM: Compiling problem (target solver=OSQP).
(CVXPY) Oct 20 06:52:59 PM: Reduction chain: CvxAttr2Constr -> Qp2SymbolicQp -> QpMatrixStuffing 

12891.666666666666

Now all solutions and status are updated into `prob1`. 

Use `prob1.status` to check the optimality

In [113]:
prob1.status

'optimal'

Optimized objective value is stored in `prob1.value`, optimized variables are stored in `x1.value`, and the Lagrange multiplier/dual value results are stored in the field `con[0].dual_value` 

**Note:**

Now we can use `print` to print out results with a nice look

In [114]:
#print optimal value and variables
print('\033[94mOptimal value: ', f'{prob1.value:.8}' +'\033[0m')
print('\033[94mOptimal variables: ', f'x1={x1.value:.5}, x2={x2.value:.5}, x3={x3.value:.5}'+'\033[0m')
print('\033[94mPrice: ', f'{con[6].dual_value:.3}'+'\033[0m')

Optimal value:  12891.667
Optimal variables:  x1=416.67, x2=83.333, x3=0.0
Price:  28.3


<font color='blue'>**Answer to Problem 2**</font>

<font color='blue'>Optimal value:  12,892

<font color='blue'>Optimal variables:  x1=417, x2=83, x3=0 so generator 1 at 417 MW, generator 2 at 83 MW, and generator 3 not running

<font color='blue'>Price:  28.3

### Problem 3 Piece-wise linear cost function (50)

A utility operates two generators with the following piece-wise linear cost curves:

**Generator 1:** The minimal generation is 50 MW and maximum is 200 MW.
* 0 MW to 100 MW: 25 \\$/MWh
* 100 MW to 150 MW: 27 \\$/MWh
* 150 MW to 200 MW: 30 \\$/MWh


**Generator 2:** The minimal generation is 30 MW and maximum is 100 MW. 
* 0 MW to 50 MW: 23 \\$/MWh
* 50 MW to 80 MW: 25 \\$/MWh
* 80 MW to 100 MW: 28 \\$/MWh


The demand to be served is 185 MW.


1) Formulate the problem into a linear proramming. Use Latex to write out your solution (refer to Problem 1 Markdown format), explain your variable definition and constraints, **do not** write code for this question. **Hint:** You should define a different variable for each segment. (20)

2) Use CVXPY to solve this problem. (20)

3) Do we need to model the segment transition logic in this case (i.e., enforce a upper segment to be zero if the lower segment is not "full" ), why? (10)

<font color='blue'>**Start of Problem 3 Part 1**</font>

<font color='blue'>**Define variables for generator 1 $p_{1}$, different variables per segment:**

$p_{11}$ for 0 MW to 100 MW segment

HOWEVER note minimum generation is 50 MW, thus

<font color='blue'>$p_{11}$ for 50 MW to 100 MW segment

<font color='blue'>$p_{12}$  for 100 MW to 150 MW segment (0 to 50 "units")

<font color='blue'>$p_{13}$  for 150 MW to 200 MW segment (0 to 50 "units")

Thus associated constaints, in UNIT form/subtract out max generation from previous segment:

<font color='blue'>$p_{11} \geq 50 $

<font color='blue'>$p_{11} \leq 100 $

<font color='blue'>$p_{12} \geq 0 $

<font color='blue'>$p_{12} \leq 50 $

<font color='blue'>$p_{13} \geq 0 $

<font color='blue'>$p_{13}  \leq 50$

With associated costs:

<font color='blue'>$c_{11} = 25 \$/MWh$

<font color='blue'>$c_{12} = 27 \$/MWh$

<font color='blue'>$c_{13} = 30 \$/MWh$

<font color='blue'>**For generator 2 $p_{2}$, different variables per segment:**

$p_{21}$ for 0 MW to 50 MW segment

HOWEVER note minimum generation is 30 MW, thus

<font color='blue'>$p_{21}$ for 30 MW to 50 MW segment

<font color='blue'>$p_{22}$  for 50 MW to 80 MW segment (0 to 30 "units")

<font color='blue'>$p_{23}$  for 80 MW to 100 MW segment (0 to 20 "units")

Thus associated constaints, in UNIT form/subtract out max generation from previous segment:

<font color='blue'>$p_{21} \geq 30 $

<font color='blue'>$p_{21} \leq 50 $

<font color='blue'>$p_{22} \geq 0 $

<font color='blue'>$p_{22} \leq 30 $

<font color='blue'>$p_{23} \geq 0 $

<font color='blue'>$p_{23}  \leq 20 $

With associated costs:

<font color='blue'>$c_{21} = 23 \$/MWh$

<font color='blue'>$c_{22} = 25 \$/MWh$

<font color='blue'>$c_{23} = 28 \$/MWh$

Total demand constraint:

<font color='blue'>$ p_{11} + p_{12} + p_{13} + p_{21} + p_{22} + p_{23} = 185 MW $

Or rearranged:

<font color='blue'>$ 185 - p_{11} - p_{12} - p_{13} - p_{21} - p_{22} - p_{23} = 0 $

<font color='blue'>**Thus objective function to be minimized is:**
$$ 25p_{11} + 27p_{12} + 30p_{13} + 23p_{21} + 25p_{22} + 28p_{23} $$

<font color='blue'>**Answer to Problem 3 Part 1: (Summarized from above)**

<font color='blue'>**Objective function to be minimized is:**
<font color='blue'>$$ 25p_{11} + 27p_{12} + 30p_{13} + 23p_{21} + 25p_{22} + 28p_{23} $$

<font color='blue'>**Subject to constraints:**

<font color='blue'>$ 185 - p_{11} - p_{12} - p_{13} - p_{21} - p_{22} - p_{23} = 0 $

<font color='blue'>$p_{11} \geq 50 $

<font color='blue'>$p_{11} \leq 100 $

<font color='blue'>$p_{12} \geq 0 $

<font color='blue'>$p_{12} \leq 50 $

<font color='blue'>$p_{13} \geq 0 $

<font color='blue'>$p_{13}  \leq 50$

<font color='blue'>$p_{21} \geq 30 $

<font color='blue'>$p_{21} \leq 50 $

<font color='blue'>$p_{22} \geq 0 $

<font color='blue'>$p_{22} \leq 30 $

<font color='blue'>$p_{23} \geq 0 $

<font color='blue'>$p_{23}  \leq 20 $

<font color='blue'>**Start of Problem 3 Part 2**</font>

In [163]:
# Define variables
p11 = cp.Variable(nonneg = True) #because won't have negative generation, not physically possible
p12 = cp.Variable(nonneg = True)
p13 = cp.Variable(nonneg = True)
p21 = cp.Variable(nonneg = True)
p22 = cp.Variable(nonneg = True)
p23 = cp.Variable(nonneg = True)

In [164]:
# Create objective function
obj3 = cp.Minimize(25*p11 + 27*p12 + 30*p13 + 23*p21 + 25*p22 + 28*p23)

In [165]:
con3 = [];

# Define constraints for generator 1
con3 += [p11 >= 50];
con3 += [p11 <= 100];
con3 += [p12 >= 0];
con3 += [p12 <= 50];
con3 += [p13 >= 0];
con3 += [p13 <= 50];
# Define constraints for generator 2
con3 += [p21 >= 30];
con3 += [p21 <= 50];
con3 += [p22 >= 0];
con3 += [p22 <= 30];
con3 += [p23 >= 0];
con3 += [p23 <= 20];
# Define demand balance constraint
con3 += [185 - p11 - p12 - p13 - p21 - p22 - p23 == 0]

In [166]:
# Solve the problem
prob3 = cp.Problem(obj3, con3)

In [167]:
# call solver to solve the problem
prob3.solve() # You may specify the solver here, 
                           # otherwise ECOS is used for LPs by default

4535.000000478926

In [168]:
prob3.status

'optimal'

In [176]:
#print optimal value and variables
print('\033[94mOptimal value: ', f'{prob3.value:.8}' +'\033[0m')
print('\033[94mOptimal variables: ', f'p11={p11.value:.5}, p12={p12.value:.5}, p13={p13.value:.5},p21={p21.value:.5},p22={p22.value:.5},p23={p23.value:.5}'+'\033[0m')
print('\033[94mPrice: ', f'{con3[-1].dual_value:.3}'+'\033[0m')

Optimal value:  4535.0
Optimal variables:  p11=100.0, p12=5.0, p13=3.8448e-08,p21=50.0,p22=30.0,p23=2.1812e-07
Price:  27.0


In [173]:
#Check: 
p11.value + p12.value + p13.value + p21.value + p22.value + p23.value

184.99999999999997

In [174]:
#Thus, generator 1 will run:
p11.value + p12.value + p13.value 

104.99999980940304

In [175]:
#Thus, generator 2 will run:
p21.value + p22.value + p23.value

80.00000019059692

<font color='blue'>**Answer to Problem 2 Part 2**</font>

<font color='blue'>**Generator 1 would total $105 MW$, generator 2 would total $80 MW$. Price would be $\$27/MWh$, which makes sense as we'd get into the second segment of generator 1 ($\$27/MWh$) and the second segment of generator 2 ($\$25/MWh$) thus generator 1 would set the price. Breakdown of values by segment printed above and summarized below.**

<font color='blue'>p11=100.0

<font color='blue'>p12=5.0

<font color='blue'>p13=0

<font color='blue'>p21=50.0

<font color='blue'>p22=30.0

<font color='blue'>p23=0

<font color='blue'>**Answer to Problem 2 Part 3**</font>

<font color='blue'>No, we don't need to model the segment transition logic in this case. Since we subtracted the generation of the previous segments from the later ones, this is already accounted for. This works because each segment has increasing cost, so the objective function will naturally minimize the later segments in favor of using the maximum possible from the earlier segments. If we hadn't defined the segment constraints in terms of "units" as opposed to cumulative generation and/or cost did not increase this way, we'd need to model the segment transition logic. 